### ISSUE TITLE
Feature: `:allow_nil` to use in conjunction with `using: SomeSerializer`, and option `serializes`

### ISSUE THREAD

**Issue by nicooga** [elixir-maru/maru_entity#14]
Reading tests I see that the only way to avoid an error when serializing a nested record that may not be present is using the `:if` option. I think is a good default for the library to raise an error when the associated record is not present. But it would also be nice to have an `allow_nil` or similar named option that served as a shortcut to achieve this behavior. This way we could go from writing this:

~~~elixir
expose :post,
 using: PostEntity,
 if: fn
   (%Comment{post: %Post{}}, _opts) -> true
   (_, _) -> false
 end
~~~

To this:
~~~elixir
expose :post, using: PostEntity, allow_nil: true
~~~

Yes I know that's a lot of of unneeded pattern matching. But I think that the more pattern match you can do the better, since you are ensuring that your serializer is not passed something it does not expect. In that sense I think it may also good idea to add a ` serializes SomeModule` top level option to specify what struct or structs this entity can accept:

~~~elixir
defmodule PostEntity do
  use Maru.entity
  serializes Post
  expose :id 
end
~~~

**Comment by falood** [comment]
I think `allow_nil: true` is useful, but I can't get the idea about `serializes SomModule`, if we have such code:
```elixir
expose :post, using: PostEntity

defmodule PostEntity do
  serializers Post
end
```
what will happen when we pass `%{post: %NotPostStruct{}}` to it?

**Comment by nicooga** [comment]
I'd expect `Entity.serialize!(map)` to raise an error if `map` is not a struct of the expected type.

~~~elixir
defmodule Dog do
  defstruct [:color, :age]
end

defmodule DogEntity do
  use Maru.Entity
  serializes Dog
end

# Option 1, Not the "elixir way" but keeps the API the same
DogEntity.serialize(%Dog{}) # returns a serialized dog
DogEntity.serialize(%NotADog{}) # raises Maru.SerializationError

# Option 2, modifies API but I think is better
DogEntity.serialize(%Dog{}) # returns {:ok, serialized_dog}
DogEntity.serialize(%NotADog{}) # returns: {:error, :invalid_module_type}
DogEntity.serialize!(%Dog{}) # returns dog
DogEntity.serialize!(%NotADog{}) # raises Maru.SerializationError
~~~

**Comment by falood** [comment]
Hmm, I don't think we need verify check like this.
In my use case, sometimes I need DogEntity.serialize for both `%Dog{}` and general map.

**Comment by nicooga** [comment]
I know not anyone needs it but it would be nice to have. You could also pass a list of modules to allow multiple modules. I think we need more opinions here. Additionally, I've noticed you cannot pass a captured function to the `:if` option:

~~~elixir
expose :some_attr, if: &some_func/2
~~~

... which also limits your options.

Anyway, there may be better things to work on at the moment.

**Comment by falood** [comment]
Yes, I'm using it to parse data from mongodb, I also think I missed some important features.
Let's do them one by one, open one issue for one feature you need, and mark important ones as critical 🙂
